pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPushUtils.java

```diff
@@ -0,0 +1,349 @@
+import org.apache.pinot.spi.ingestion.batch.spec.PinotClusterSpec;
+import org.apache.pinot.spi.ingestion.batch.spec.SegmentGenerationJobSpec;
+import org.apache.pinot.spi.utils.retry.AttemptsExceededException;
+import org.apache.pinot.spi.utils.retry.RetriableOperationException;
+import org.apache.pinot.spi.utils.retry.RetryPolicies;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class SegmentPushUtils implements Serializable {
+  private SegmentPushUtils() {
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPushUtils.class);
+  private static final FileUploadDownloadClient FILE_UPLOAD_DOWNLOAD_CLIENT = new FileUploadDownloadClient();
+
+  public static URI generateSegmentTarURI(URI dirURI, URI fileURI, String prefix, String suffix) {
+    if (StringUtils.isEmpty(prefix) && StringUtils.isEmpty(suffix)) {
+      // In case the FS doesn't provide scheme or host, will fill it up from dirURI.
+      String scheme = fileURI.getScheme();
+      if (StringUtils.isEmpty(fileURI.getScheme())) {
+        scheme = dirURI.getScheme();
+      }
+      String host = fileURI.getHost();
+      if (StringUtils.isEmpty(fileURI.getHost())) {
+        host = dirURI.getHost();
+      }
+      int port = fileURI.getPort();
+      if (port < 0) {
+        port = dirURI.getPort();
+      }
+      try {
+        return new URI(scheme, fileURI.getUserInfo(), host, port, fileURI.getPath(), fileURI.getQuery(),
+            fileURI.getFragment());
+      } catch (URISyntaxException e) {
+        LOGGER.warn("Unable to generate push uri based from dir URI: {} and file URI: {}, directly return file URI.",
+            dirURI, fileURI);
+        return fileURI;
+      }
+    }
+    return URI.create(String.format("%s%s%s", prefix, fileURI.getRawPath(), suffix));
+  }
+
+  public static void pushSegments(SegmentGenerationJobSpec spec, PinotFS fileSystem, List<String> tarFilePaths)
+      throws RetriableOperationException, AttemptsExceededException {
+    String tableName = spec.getTableSpec().getTableName();
+    boolean cleanUpOutputDir = spec.isCleanUpOutputDir();
+    LOGGER.info("Start pushing segments: {}... to locations: {} for table {}",
+        Arrays.toString(tarFilePaths.subList(0, Math.min(5, tarFilePaths.size())).toArray()),
+        Arrays.toString(spec.getPinotClusterSpecs()), tableName);
+    for (String tarFilePath : tarFilePaths) {
+      URI tarFileURI = URI.create(tarFilePath);
+      File tarFile = new File(tarFilePath);
+      String fileName = tarFile.getName();
+      Preconditions.checkArgument(fileName.endsWith(Constants.TAR_GZ_FILE_EXT));
+      String segmentName = fileName.substring(0, fileName.length() - Constants.TAR_GZ_FILE_EXT.length());
+      for (PinotClusterSpec pinotClusterSpec : spec.getPinotClusterSpecs()) {
+        URI controllerURI;
+        try {
+          controllerURI = new URI(pinotClusterSpec.getControllerURI());
+        } catch (URISyntaxException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)